aip/0166.md

```diff
@@ -0,0 +1,78 @@
+---
+aip:
+  id: 166
+  state: reviewing
+  created: 2020-06-24
+permalink: /166
+redirect_from:
+  - /0166
+---
+
+# Hierarchically-derived resources
+
+Services occasionally have resources where the resource's effective values are
+determined by consulting both configuration on the resource as well as
+configuration elsewhere, such as a parent in the hierarchy, and "flattening"
+them together. This is most common for configuration resources, which could be
+applied against a resource or a container.
+
+In these situations, users need to distinguish between properties applied to
+the resource directly and properties that are inherited.
+
+## Guidance
+
+Services **should** return only field values on the resource when responding to
+`Get` and `List` methods by default.
+
+Additionally, services **may** add a field to `Get` and `List` requests that
+allows the user to opt-in to receiving the effective configuration instead:
+
+```proto
+message GetBookRequest {
+  // The name of the book.
+  string name = 1 [(google.api.resource_reference) = {
+    type: "library.googleapis.com/Book"
+  }];
+
+  // If true, values that are set on other objects in the hierachy and
+  // inherited by the book will be included.
+  // If false (default), only fields directly set on the book are included.
+  bool effective = 2;
+}
+```
+
+Services **must not** add this field for cases where determining the effective
+values is expensive.
+
+### Expensive computations
+
+On rare occasion, the effective values for a resource can be expensive to
+compute and not referenced often. Additionally, a service might need to charge
+for the query.
+
+In these situations, the service **may** provide a specialized method:
+
+```proto
+rpc ComputeEffectiveBook(ComputeEffectiveBookRequest) returns (Book) {
+  option (google.api.http) = {
+    post: "/v1/{name=publishers/*/books/*}:computeEffective"
+    body: "*"
+  };
+}
+
+message ComputeEffectiveBookRequest {
+  // The name of the book.
+  string name = 1 [(google.api.resource_reference) = {
+    type: "library.googleapis.com/Book"
+  }];
+}
+```
+
+- The method **should** return the resource instead of a custom response
```

Comment:
When returning the resource which is a composite of inherited resources/policies, what should the name field be? Blank or the resource that was queried?

Reply:
The resource queried.